core/lib/zksync_core/src/lib.rs

```diff
@@ -402,6 +402,10 @@ pub async fn initialize_components(
                 .get_or_init()
                 .await
                 .context("gas_adjuster.get_or_init()")?;
+            let batch_fee_input_provider = Arc::new(MainNodeFeeInputProvider::new(
+                bounded_gas_adjuster,
+                FeeModelConfig::from_state_keeper_config(&state_keeper_config),
+            ));
```

Comment:
do you really need to construct FeeInput provider 3 times?

Reply:
Just like gas adjuster, which is also created via `get_or_init` 3 times (so that it is only instantiated on demand).
There exists an open question of whether we want to merge `GasAdjuster` (or `GasAdjusterSingleton`) and `MainNodeFeeInputProvider`, but I think it's out of scope for this PR (as it's about getting rid of abstraction).

Reply:
yeah, but get or init returns the only instance. 
but here you have 3 instances. 
could it lead to unnecessary requests to the server?

Reply:
No